// @gluestack-ui/themed/build/components/core/Menu/styled-components/Item.web.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Item = void 0;
const styled_1 = require("../../styled");
const html_elements_1 = require("@expo/html-elements");
exports.Item = (0, styled_1.styled)(html_elements_1.LI, {
    'p': '$3',
    'flexDirection': 'row',
    'alignItems': 'center',
    ':hover': {
        bg: '$backgroundLight100',
    },
    ':disabled': {
        'opacity': 0.4,
        // @ts-ignore
        'cursor': 'not-allowed',
        ':focus': {
            bg: 'transparent',
        },
        '_dark': {
            ':focus': {
                bg: 'transparent',
            },
        },
    },
    ':active': {
        bg: '$backgroundLight200',
    },
    ':focus': {
        bg: '$backgroundLight100',
        // @ts-ignore
        outlineWidth: '$0',
        outlineStyle: 'none',
    },
    '_dark': {
        ':hover': {
            bg: '$backgroundDark800',
        },
        ':active': {
            bg: '$backgroundDark700',
        },
        ':focus': {
            bg: '$backgroundDark800',
        },
    },
    ':focusVisible': {
        // @ts-ignore
        outlineWidth: '$0.5',
        outlineColor: '$primary700',
        outlineStyle: 'solid',
        _dark: {
            outlineColor: '$primary300',
        },
    },
    'cursor': 'pointer',
}, {
    descendantStyle: ['_text'],
});
