// @gluestack-ui/themed/build/components/core/Actionsheet/styled-components/Item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const styled_1 = require("../../styled");
const react_native_1 = require("react-native");
exports.default = (0, styled_1.styled)(react_native_1.Pressable, {
    p: "$3",
    flexDirection: "row",
    alignItems: "center",
    rounded: "$sm",
    w: "100%",
    ":disabled": {
        opacity: 0.4,
        // @ts-ignore
        _web: {
            // @ts-ignore
            pointerEvents: "all !important",
            cursor: "not-allowed",
        },
    },
    ":hover": {
        bg: "$backgroundLight50",
    },
    ":active": {
        bg: "$backgroundLight100",
    },
    ":focus": {
        bg: "$backgroundLight100",
    },
    _dark: {
        ":hover": {
            bg: "$backgroundDark800",
        },
        ":active": {
            bg: "$backgroundDark700",
        },
        ":focus": {
            bg: "$backgroundDark700",
        },
    },
    _web: {
        ":focusVisible": {
            bg: "$backgroundLight100",
            _dark: {
                bg: "$backgroundDark700",
            },
        },
    },
}, {
    descendantStyle: ["_text"],
    DEBUG: "ACTIONSHEET_ITEM",
});
