// @gluestack-ui/themed/build/components/core/FormControl/styled-components/Root.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
const react_native_1 = require("react-native");
const styled_1 = require("../../styled");
exports.default = (0, styled_1.styled)(react_native_1.View, {
    flexDirection: 'column',
    variants: {
        size: {
            sm: {
                _labelText: {
                    fontSize: '$sm',
                },
                _labelAstrick: {
                    fontSize: '$sm',
                },
                _helperText: {
                    fontSize: '$xs',
                },
                _errorText: {
                    fontSize: '$xs',
                },
            },
            md: {
                _labelText: {
                    fontSize: '$md',
                },
                _labelAstrick: {
                    fontSize: '$md',
                },
                _helperText: {
                    fontSize: '$sm',
                },
                _errorText: {
                    fontSize: '$sm',
                },
            },
            lg: {
                _labelText: {
                    fontSize: '$lg',
                },
                _labelAstrick: {
                    fontSize: '$lg',
                },
                _helperText: {
                    fontSize: '$md',
                },
                _errorText: {
                    fontSize: '$md',
                },
            },
        },
    },
    defaultProps: {
        size: 'md',
    },
}, {
    descendantStyle: [
        '_labelText',
        '_helperText',
        '_errorText',
        '_labelAstrick',
    ],
});
