// @gluestack-ui/themed/build/components/core/FormControl/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormControlHelperText = exports.FormControlHelper = exports.FormControlLabelAstrick = exports.FormControlLabelText = exports.FormControlLabel = exports.FormControlErrorIcon = exports.FormControlErrorText = exports.FormControlError = exports.FormControl = void 0;
const form_control_1 = require("@gluestack-ui/form-control");
const styled_components_1 = require("./styled-components");
exports.FormControl = (0, form_control_1.createFormControl)({
    Root: styled_components_1.Root,
    Error: styled_components_1.Error,
    ErrorText: styled_components_1.ErrorText,
    ErrorIcon: styled_components_1.ErrorIcon,
    Label: styled_components_1.Label,
    LabelText: styled_components_1.LabelText,
    LabelAstrick: styled_components_1.LabelAstrick,
    Helper: styled_components_1.Helper,
    HelperText: styled_components_1.HelperText,
});
exports.FormControlError = exports.FormControl.Error;
exports.FormControlErrorText = exports.FormControl.Error.Text;
exports.FormControlErrorIcon = exports.FormControl.Error.Icon;
exports.FormControlLabel = exports.FormControl.Label;
exports.FormControlLabelText = exports.FormControl.Label.Text;
exports.FormControlLabelAstrick = exports.FormControl.Label.Astrick;
exports.FormControlHelper = exports.FormControl.Helper;
exports.FormControlHelperText = exports.FormControl.Helper.Text;
