// @gluestack-ui/themed/build/components/core/Actionsheet/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionsheetIcon = exports.ActionsheetSectionHeaderText = exports.ActionsheetSectionList = exports.ActionsheetFlatList = exports.ActionsheetVirtualizedList = exports.ActionsheetScrollView = exports.ActionsheetBackdrop = exports.ActionsheetDragIndicatorWrapper = exports.ActionsheetDragIndicator = exports.ActionsheetItemText = exports.ActionsheetItem = exports.ActionsheetContent = exports.Actionsheet = void 0;
const actionsheet_1 = require("@gluestack-ui/actionsheet");
const styled_1 = require("../styled");
const styled_components_1 = require("./styled-components");
exports.Actionsheet = (0, actionsheet_1.createActionsheet)({
    Root: styled_components_1.Root,
    Content: styled_components_1.Content,
    Item: styled_components_1.Item,
    ItemText: styled_components_1.ItemText,
    DragIndicator: styled_components_1.DragIndicator,
    IndicatorWrapper: styled_components_1.IndicatorWrapper,
    Backdrop: styled_components_1.Backdrop,
    ScrollView: styled_components_1.ScrollView,
    VirtualizedList: styled_components_1.VirtualizedList,
    FlatList: styled_components_1.FlatList,
    SectionList: styled_components_1.SectionList,
    SectionHeaderText: styled_components_1.SectionHeaderText,
    Icon: styled_components_1.Icon,
    //@ts-ignore
    AnimatePresence: styled_1.styled.Component,
});
exports.ActionsheetContent = exports.Actionsheet.Content;
exports.ActionsheetItem = exports.Actionsheet.Item;
exports.ActionsheetItemText = exports.Actionsheet.ItemText;
exports.ActionsheetDragIndicator = exports.Actionsheet.DragIndicator;
exports.ActionsheetDragIndicatorWrapper = exports.Actionsheet.DragIndicatorWrapper;
exports.ActionsheetBackdrop = exports.Actionsheet.Backdrop;
exports.ActionsheetScrollView = exports.Actionsheet.ScrollView;
exports.ActionsheetVirtualizedList = exports.Actionsheet.VirtualizedList;
exports.ActionsheetFlatList = exports.Actionsheet.FlatList;
exports.ActionsheetSectionList = exports.Actionsheet.SectionList;
exports.ActionsheetSectionHeaderText = exports.Actionsheet.SectionHeaderText;
exports.ActionsheetIcon = exports.Actionsheet.Icon;
