// @gluestack-ui/themed/build/components/core/Popover/styled-components/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Arrow = exports.Backdrop = exports.Body = exports.Footer = exports.Header = exports.CloseButton = exports.Content = exports.Root = void 0;
var Root_1 = require("./Root");
Object.defineProperty(exports, "Root", { enumerable: true, get: function () { return __importDefault(Root_1).default; } });
var Content_1 = require("./Content");
Object.defineProperty(exports, "Content", { enumerable: true, get: function () { return __importDefault(Content_1).default; } });
var CloseButton_1 = require("./CloseButton");
Object.defineProperty(exports, "CloseButton", { enumerable: true, get: function () { return __importDefault(CloseButton_1).default; } });
var Header_1 = require("./Header");
Object.defineProperty(exports, "Header", { enumerable: true, get: function () { return __importDefault(Header_1).default; } });
var Footer_1 = require("./Footer");
Object.defineProperty(exports, "Footer", { enumerable: true, get: function () { return __importDefault(Footer_1).default; } });
var Body_1 = require("./Body");
Object.defineProperty(exports, "Body", { enumerable: true, get: function () { return __importDefault(Body_1).default; } });
var Backdrop_1 = require("./Backdrop");
Object.defineProperty(exports, "Backdrop", { enumerable: true, get: function () { return __importDefault(Backdrop_1).default; } });
var Arrow_1 = require("./Arrow");
Object.defineProperty(exports, "Arrow", { enumerable: true, get: function () { return __importDefault(Arrow_1).default; } });
