// @gluestack-ui/themed/build/components/core/Checkbox/styled-components/Label.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const react_native_1 = require("react-native");
const styled_1 = require("../../styled");
exports.default = (0, styled_1.styled)(react_native_1.Text, {
    'color': '$textLight600',
    ':checked': {
        color: '$textLight900',
    },
    ':hover': {
        'color': '$textLight900',
        ':checked': {
            'color': '$textLight900',
            ':disabled': {
                color: '$textLight900',
            },
        },
        ':disabled': {
            color: '$textLight600',
        },
    },
    ':active': {
        'color': '$textLight900',
        ':checked': {
            color: '$textLight900',
        },
    },
    ':disabled': {
        opacity: 0.4,
    },
    '_web': {
        MozUserSelect: 'none',
        WebkitUserSelect: 'none',
        msUserSelect: 'none',
    },
    '_dark': {
        'color': '$textDark400',
        ':checked': {
            color: '$textDark100',
        },
        ':hover': {
            'color': '$textDark100',
            ':checked': {
                'color': '$textDark100',
                ':disabled': {
                    color: '$textDark100',
                },
            },
        },
        ':disabled': {
            color: '$textDark100',
        },
        ':active': {
            'color': '$textDark100',
            ':checked': {
                color: '$textDark100',
            },
        },
    },
}, {
    ancestorStyle: ['_text'],
});
