// @gluestack-ui/themed/build/components/core/Link/styled-components/Root.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const styled_1 = require("../../styled");
const react_native_1 = require("react-native");
exports.default = (0, styled_1.styled)(react_native_1.Pressable, {
    'outlineWidth': 0,
    ':focusVisible': {
        outlineWidth: 2,
        outlineColor: '$primary700',
        outlineStyle: 'solid',
        _dark: {
            // @ts-ignore
            outlineColor: '$primary400',
        },
    },
    '_text': {
        'fontWeight': '$normal',
        'textDecorationLine': 'underline',
        'color': '$info700',
        ':hover': {
            color: '$info600',
            textDecorationLine: 'none',
        },
        ':active': {
            color: '$info700',
        },
        ':disabled': {
            opacity: 0.4,
        },
        '_dark': {
            'color': '$info300',
            ':hover': {
                color: '$info400',
                textDecorationLine: 'none',
            },
            ':active': {
                color: '$info300',
            },
            ':disabled': {
                opacity: 0.4,
            },
        },
    },
    '_web': {
        ':disabled': {
            cursor: 'not-allowed',
        },
    },
}, {
    descendantStyle: ['_text'],
});
