// @gluestack-ui/themed/build/components/core/Tabs/styled-components/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TabIcon = exports.TabTitle = exports.TabPanels = exports.TabPanel = exports.TabList = exports.Tab = exports.Root = void 0;
var Root_1 = require("./Root");
Object.defineProperty(exports, "Root", { enumerable: true, get: function () { return __importDefault(Root_1).default; } });
var Tab_1 = require("./Tab");
Object.defineProperty(exports, "Tab", { enumerable: true, get: function () { return __importDefault(Tab_1).default; } });
var TabList_1 = require("./TabList");
Object.defineProperty(exports, "TabList", { enumerable: true, get: function () { return __importDefault(TabList_1).default; } });
var TabPanel_1 = require("./TabPanel");
Object.defineProperty(exports, "TabPanel", { enumerable: true, get: function () { return __importDefault(TabPanel_1).default; } });
var TabPanels_1 = require("./TabPanels");
Object.defineProperty(exports, "TabPanels", { enumerable: true, get: function () { return __importDefault(TabPanels_1).default; } });
var TabTitle_1 = require("./TabTitle");
Object.defineProperty(exports, "TabTitle", { enumerable: true, get: function () { return __importDefault(TabTitle_1).default; } });
var TabIcon_1 = require("./TabIcon");
Object.defineProperty(exports, "TabIcon", { enumerable: true, get: function () { return __importDefault(TabIcon_1).default; } });
