// @gluestack-ui/themed/build/components/core/AlertDialog/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertDialogBackdrop = exports.AlertDialogBody = exports.AlertDialogFooter = exports.AlertDialogHeader = exports.AlertDialogCloseButton = exports.AlertDialogContent = exports.AlertDialog = void 0;
const alert_dialog_1 = require("@gluestack-ui/alert-dialog");
const styled_components_1 = require("./styled-components");
const styled_1 = require("../styled");
exports.AlertDialog = (0, alert_dialog_1.createAlertDialog)({
    Root: styled_components_1.Root,
    Content: styled_components_1.Content,
    CloseButton: styled_components_1.CloseButton,
    Header: styled_components_1.Header,
    Footer: styled_components_1.Footer,
    Body: styled_components_1.Body,
    Backdrop: styled_components_1.Backdrop,
    //@ts-ignore
    AnimatePresence: styled_1.styled.Component,
});
exports.AlertDialogContent = exports.AlertDialog.Content;
exports.AlertDialogCloseButton = exports.AlertDialog.CloseButton;
exports.AlertDialogHeader = exports.AlertDialog.Header;
exports.AlertDialogFooter = exports.AlertDialog.Footer;
exports.AlertDialogBody = exports.AlertDialog.Body;
exports.AlertDialogBackdrop = exports.AlertDialog.Backdrop;
