// @gluestack-ui/themed/build/components/core/Modal/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModalBackdrop = exports.ModalBody = exports.ModalFooter = exports.ModalHeader = exports.ModalCloseButton = exports.ModalContent = exports.Modal = void 0;
const modal_1 = require("@gluestack-ui/modal");
const styled_components_1 = require("./styled-components");
const styled_1 = require("../styled");
exports.Modal = (0, modal_1.createModal)({
    Root: styled_components_1.Root,
    Content: styled_components_1.Content,
    CloseButton: styled_components_1.CloseButton,
    Header: styled_components_1.Header,
    Footer: styled_components_1.Footer,
    Body: styled_components_1.Body,
    Backdrop: styled_components_1.Backdrop,
    //@ts-ignore
    AnimatePresence: styled_1.styled.Component,
});
exports.ModalContent = exports.Modal.Content;
exports.ModalCloseButton = exports.Modal.CloseButton;
exports.ModalHeader = exports.Modal.Header;
exports.ModalFooter = exports.Modal.Footer;
exports.ModalBody = exports.Modal.Body;
exports.ModalBackdrop = exports.Modal.Backdrop;
