// @gluestack-ui/themed/build/components/core/Select/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectSectionHeaderText = exports.SelectSectionList = exports.SelectFlatList = exports.SelectVirtualizedList = exports.SelectScrollView = exports.SelectItemText = exports.SelectItem = exports.SelectDragIndicatorWrapper = exports.SelectDragIndicator = exports.SelectContent = exports.SelectBackdrop = exports.SelectPortal = exports.SelectIcon = exports.SelectInput = exports.SelectTrigger = exports.Select = void 0;
const select_1 = require("@gluestack-ui/select");
const actionsheet_1 = require("@gluestack-ui/actionsheet");
const styled_1 = require("../styled");
const styled_components_actionsheet_1 = require("./styled-components-actionsheet");
const styled_components_1 = require("./styled-components");
const Actionsheet = (0, actionsheet_1.createActionsheet)({
    Root: styled_components_actionsheet_1.Root,
    Backdrop: styled_components_actionsheet_1.Backdrop,
    Content: styled_components_actionsheet_1.Content,
    DragIndicator: styled_components_actionsheet_1.DragIndicator,
    IndicatorWrapper: styled_components_actionsheet_1.IndicatorWrapper,
    Item: styled_components_actionsheet_1.Item,
    ItemText: styled_components_actionsheet_1.ItemText,
    Icon: styled_components_actionsheet_1.Icon,
    ScrollView: styled_components_actionsheet_1.ScrollView,
    VirtualizedList: styled_components_actionsheet_1.VirtualizedList,
    FlatList: styled_components_actionsheet_1.FlatList,
    SectionList: styled_components_actionsheet_1.SectionList,
    SectionHeaderText: styled_components_actionsheet_1.SectionHeaderText,
    // @ts-ignore
    AnimatePresence: styled_1.styled.Component,
});
exports.Select = (0, select_1.createSelect)({
    Root: styled_components_1.Root,
    Trigger: styled_components_1.Trigger,
    Input: styled_components_1.Input,
    Icon: styled_components_1.Icon,
}, {
    Portal: Actionsheet,
    Backdrop: Actionsheet.Backdrop,
    Content: Actionsheet.Content,
    DragIndicator: Actionsheet.DragIndicator,
    DragIndicatorWrapper: Actionsheet.DragIndicatorWrapper,
    Item: Actionsheet.Item,
    ItemText: Actionsheet.ItemText,
    ScrollView: Actionsheet.ScrollView,
    VirtualizedList: Actionsheet.VirtualizedList,
    FlatList: Actionsheet.FlatList,
    SectionList: Actionsheet.SectionList,
    SectionHeaderText: Actionsheet.SectionHeaderText,
});
exports.SelectTrigger = exports.Select.Trigger;
exports.SelectInput = exports.Select.Input;
exports.SelectIcon = exports.Select.Icon;
exports.SelectPortal = exports.Select.Portal;
exports.SelectBackdrop = exports.Select.Backdrop;
exports.SelectContent = exports.Select.Content;
exports.SelectDragIndicator = exports.Select.DragIndicator;
exports.SelectDragIndicatorWrapper = exports.Select.DragIndicatorWrapper;
exports.SelectItem = exports.Select.Item;
exports.SelectItemText = exports.Select.ItemText;
exports.SelectScrollView = exports.Select.ScrollView;
exports.SelectVirtualizedList = exports.Select.VirtualizedList;
exports.SelectFlatList = exports.Select.FlatList;
exports.SelectSectionList = exports.Select.SectionList;
exports.SelectSectionHeaderText = exports.Select.SectionHeaderText;
