// @gluestack-ui/themed/build/components/core/Popover/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PopoverCloseButton = exports.PopoverBackdrop = exports.PopoverBody = exports.PopoverFooter = exports.PopoverHeader = exports.PopoverContent = exports.PopoverArrow = exports.Popover = void 0;
const popover_1 = require("@gluestack-ui/popover");
const styled_1 = require("../styled");
const styled_components_1 = require("./styled-components");
exports.Popover = (0, popover_1.createPopover)({
    Root: styled_components_1.Root,
    Arrow: styled_components_1.Arrow,
    Content: styled_components_1.Content,
    Header: styled_components_1.Header,
    Footer: styled_components_1.Footer,
    Body: styled_components_1.Body,
    Backdrop: styled_components_1.Backdrop,
    CloseButton: styled_components_1.CloseButton,
    //@ts-ignore
    AnimatePresence: styled_1.styled.Component,
});
exports.PopoverArrow = exports.Popover.Arrow;
exports.PopoverContent = exports.Popover.Content;
exports.PopoverHeader = exports.Popover.Header;
exports.PopoverFooter = exports.Popover.Footer;
exports.PopoverBody = exports.Popover.Body;
exports.PopoverBackdrop = exports.Popover.Backdrop;
exports.PopoverCloseButton = exports.Popover.CloseButton;
