// @gluestack-style/react/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  styled: true,
  verboseStyled: true,
  StyledProvider: true,
  useStyled: true,
  styledToStyledResolved: true,
  styledResolvedToOrderedSXResolved: true,
  flush: true,
  convertStyledToStyledVerbosed: true,
  convertSxToSxVerbosed: true,
  createStyled: true,
  createGlobalStylesWeb: true,
  AsForwarder: true,
  AddCssTokenVariables: true,
  FontResolver: true,
  INTERNAL_updateCSSStyleInOrderedResolved: true,
  createConfig: true
};
Object.defineProperty(exports, "AddCssTokenVariables", {
  enumerable: true,
  get: function () {
    return _plugins.AddCssTokenVariables;
  }
});
Object.defineProperty(exports, "AsForwarder", {
  enumerable: true,
  get: function () {
    return _AsForwarder.AsForwarder;
  }
});
Object.defineProperty(exports, "FontResolver", {
  enumerable: true,
  get: function () {
    return _plugins.FontResolver;
  }
});
Object.defineProperty(exports, "INTERNAL_updateCSSStyleInOrderedResolved", {
  enumerable: true,
  get: function () {
    return _updateCSSStyleInOrderedResolved.INTERNAL_updateCSSStyleInOrderedResolved;
  }
});
Object.defineProperty(exports, "StyledProvider", {
  enumerable: true,
  get: function () {
    return _StyledProvider.StyledProvider;
  }
});
Object.defineProperty(exports, "convertStyledToStyledVerbosed", {
  enumerable: true,
  get: function () {
    return _convertSxToSxVerbosed.convertStyledToStyledVerbosed;
  }
});
Object.defineProperty(exports, "convertSxToSxVerbosed", {
  enumerable: true,
  get: function () {
    return _convertSxToSxVerbosed.convertSxToSxVerbosed;
  }
});
Object.defineProperty(exports, "createConfig", {
  enumerable: true,
  get: function () {
    return _createConfig.createConfig;
  }
});
Object.defineProperty(exports, "createGlobalStylesWeb", {
  enumerable: true,
  get: function () {
    return _createGlobalStylesWeb.createGlobalStylesWeb;
  }
});
Object.defineProperty(exports, "createStyled", {
  enumerable: true,
  get: function () {
    return _createStyled.createStyled;
  }
});
Object.defineProperty(exports, "flush", {
  enumerable: true,
  get: function () {
    return _injectInStyle.flush;
  }
});
Object.defineProperty(exports, "styled", {
  enumerable: true,
  get: function () {
    return _styled.styled;
  }
});
Object.defineProperty(exports, "styledResolvedToOrderedSXResolved", {
  enumerable: true,
  get: function () {
    return _resolver.styledResolvedToOrderedSXResolved;
  }
});
Object.defineProperty(exports, "styledToStyledResolved", {
  enumerable: true,
  get: function () {
    return _resolver.styledToStyledResolved;
  }
});
Object.defineProperty(exports, "useStyled", {
  enumerable: true,
  get: function () {
    return _StyledProvider.useStyled;
  }
});
Object.defineProperty(exports, "verboseStyled", {
  enumerable: true,
  get: function () {
    return _styled.verboseStyled;
  }
});
var _styled = require("./styled");
var _StyledProvider = require("./StyledProvider");
var _resolver = require("./resolver");
var _injectInStyle = require("./injectInStyle");
var _convertSxToSxVerbosed = require("./convertSxToSxVerbosed");
var _createStyled = require("./createStyled");
var _createGlobalStylesWeb = require("./createGlobalStylesWeb");
var _AsForwarder = require("./AsForwarder");
var _plugins = require("./plugins");
var _updateCSSStyleInOrderedResolved = require("./updateCSSStyleInOrderedResolved");
var _createConfig = require("./createConfig");
var _core = require("./core");
Object.keys(_core).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _core[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _core[key];
    }
  });
});
