// @gluestack-ui/themed/build/components/core/Heading/styled-components/Root.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const styled_1 = require("../../styled");
const html_elements_1 = require("@expo/html-elements");
exports.default = (0, styled_1.styled)(html_elements_1.H3, {
    color: '$textLight900',
    letterSpacing: '$sm',
    fontWeight: '$bold',
    fontFamily: '$heading',
    // Overrides expo-html default styling
    marginVertical: 0,
    _dark: {
        color: '$textDark50',
    },
    variants: {
        isTruncated: {
            true: {
                props: {
                    // @ts-ignore
                    numberOfLines: 1,
                    ellipsizeMode: 'tail',
                },
            },
        },
        bold: {
            true: {
                fontWeight: '$bold',
            },
        },
        underline: {
            true: {
                textDecorationLine: 'underline',
            },
        },
        strikeThrough: {
            true: {
                textDecorationLine: 'line-through',
            },
        },
        sub: {
            true: {
                fontSize: '$xs',
                lineHeight: '$xs',
            },
        },
        italic: {
            true: {
                fontStyle: 'italic',
            },
        },
        highlight: {
            true: {
                bg: '$yellow500',
            },
        },
        size: {
            '5xl': {
                //@ts-ignore
                props: { as: html_elements_1.H1 },
                fontSize: '$6xl',
                lineHeight: '$7xl',
            },
            '4xl': {
                //@ts-ignore
                props: { as: html_elements_1.H1 },
                fontSize: '$5xl',
                lineHeight: '$6xl',
            },
            '3xl': {
                //@ts-ignore
                props: { as: html_elements_1.H1 },
                fontSize: '$4xl',
                lineHeight: '$5xl',
            },
            '2xl': {
                //@ts-ignore
                props: { as: html_elements_1.H2 },
                fontSize: '$3xl',
                lineHeight: '$3xl',
            },
            'xl': {
                //@ts-ignore
                props: { as: html_elements_1.H3 },
                fontSize: '$2xl',
                lineHeight: '$3xl',
            },
            'lg': {
                //@ts-ignore
                props: { as: html_elements_1.H4 },
                fontSize: '$xl',
                lineHeight: '$2xl',
            },
            'md': {
                //@ts-ignore
                props: { as: html_elements_1.H5 },
                fontSize: '$lg',
                lineHeight: '$lg',
            },
            'sm': {
                //@ts-ignore
                props: { as: html_elements_1.H6 },
                fontSize: '$md',
                lineHeight: '$lg',
            },
            'xs': {
                //@ts-ignore
                props: { as: html_elements_1.H6 },
                fontSize: '$sm',
                lineHeight: '$xs',
            },
        },
    },
    defaultProps: {
        size: 'lg',
    },
});
